### ISSUE TITLE
Copy of the whole client connection structure at a request necessary to save the connection?

### ISSUE THREAD

**Issue by Thalak** [h2o/h2o#481]
Hi,
I want to hold the request by a client when I get a request in my callback function:

int HttpAppServer::OnRequest(h2o_handler_t *self, h2o_req_t *req)
{
      ...
     parsedUrl();
     ...
     if(m_pCallback != NULL)
     {
        m_pCallback->AppRequestReceived(parsedUrl);
     }
     ...
    //copy 'req' to my own structure before leave the function(?)
}

Because I have to call in this callback above a other own callback function (AppRequestReceived() ) to a other thread who is offering the data for the request. While this time (less mseconds e.g.) the connection to the client keeps open. 

How can I save the open client-connection until my other thread has processed the data for the response?

It is necessary to copy the whole h2o_req_t *req structure to save the connection to the client or can you tell me which parameters of the h2o_req_t structure are only necessary to save?

Thanks

**Comment by kazuho** [comment]
The only requirement of `on_req` callback is to return 0 if it decides to handle the request, or -1 if it refuses to do so.  Applications can either call `h2o_start_response` (succeeded by one or more calls to `h2o_send`) either within the `on_req` callback (i.e. handle the HTTP request synchronously) or after the `on_req` callback returned zero (i.e. asynchronous request handling).  One thing that must be noted in implementing asynchronous request handling is that you must call `h2o_start_response` and `h2o_send` functions within the same thread that called the `on_req` callback.

In your case, you need to do the following:
* in `on_req`, call `AppRequestReceived` to start a thread dedicated to handling the request, and then return zero
* when your thread finishes building the response, send the response back to the thread that called the `on_req` (either by using the functions defined in `h2o/multithread.h` or by using [`uv_async_t`](http://docs.libuv.org/en/v1.x/async.html) under the premise that you are using libh2o bound to libuv), and within the original thread call `h2o_start_response` and `h2o_send`

**Comment by Thalak** [comment]
Thank you for your good explanation of the response asynchronously.    :)

I have to build the response and send this response back to the thread who called the on_req function. I use right now 

      if (uv_run(loop, UV_RUN_DEFAULT) != 0)  {
           bOk = false;
      }

This thread will only come back if a request by a client is incoming. Therefore I should use the uv_async_t async handle, what you told me.

There is still on question open:
After my understanding I have to build the response (req)structure new, similar to:

        h2o_req_t *reqResponse;
        reqResponse = ( h2o_req_t  *)malloc(sizeof(struct  h2o_req_t ));
        h2o_iovec_t body[2];
        body[0] = h2o_strdup(&req->pool, "data1\n", SIZE_MAX);      
        body[1] = h2o_strdup(&req->pool, "data2\n", SIZE_MAX);
        req->res.reason = "OK";
        req->res.status = 200;
        ...
        h2o_start_response(reqResponse, &generator);
        h2o_send(reqResponse, body, 2, 1);
        
But I need still the handle to the connection by the requested client who is still waiting of the response. 

Which parameter, connection-id e.g., have I still to set for the response-structure?

I have to save some parameters of the connection in the on_req callback call before I leave this callback with return 0.  And if I build the response new, I need the connection handles again to address the client who is still waiting for response. 

But which handles I have to save ?

Thalak

**Comment by kazuho** [comment]
> After my understanding I have to build the response (req)structure new,

No you don't need to.  You can read/write the `h2o_req_t` from your worker thread until you call `h2o_start_response`, since the worker thread would be the only thread that would touch the structure during the time.

**Comment by Thalak** [comment]
Ahh. I have to save the request-struct pointer similar to:
 (the example only as principle explaining)

     static h2o_req_t *_reqSave;

     nt HttpAppServer::OnRequest(h2o_handler_t *self, h2o_req_t *req)
    {
           ...
           _reqSave = req;
           ...
           parsedUrl();
           ...
           if(m_pCallback != NULL)
           {
              m_pCallback->AppRequestReceived(parsedUrl);
           }
           ...
           ...
           return 0;
    }
.

Because a second callback by a other client will come before I have answered the first request by the first client. And when I have build the request struct and hand this over to the on_req thread I have to use the saved _reqSave-pointer to request the first client  and it working fine.   :)

**Comment by Thalak** [comment]
Hi,
if a save the req-struct and build the data for response in a other thread and call from the other thread the response to the client

     h2o_iovec_t body[1];
     body[0] = h2o_strdup(&_req2->pool, "Hello World\n", SIZE_MAX);
    reqResponse->_reqSave .reason = "OK";
    rereqResponseq->_reqSave.status = 200;
    h2o_start_response(_reqSave , &generator);
    h2o_send(_reqSave , body, 2, 1);

it seems to be work. There was no trouble found because of async all of the request and response.
Or can some conflicts be occur later?

Regards,
Thalak

**Comment by Thalak** [comment]
Yes. Your are right. With the async call h2o_start_response and h2o_send I got timing trouble with the h2o-thread. I have used your link 

http://docs.libuv.org/en/v1.x/async.html

"Async handles allow the user to “wakeup” the event loop and get a callback called from another thread."

and now with sync call of request and response it's working fine.   :)

and therefore I could
